[docs] Migrate filter plugins (#139)
moodle/devdocs: docs/apis/plugintypes/filter/index.md

---
title: Filter plugins
tags:
  - Filter
  - Plugins
---

<!-- cspell:ignore filtername -->
<!-- cspell:ignore strpos -->
<!-- cspell:ignore localconfig -->

import {
    Since,
    CodeBlock
} from '@site/src/components';

import {
    DbAccessPHP,
    Lang,
    Lib,
    VersionPHP,
} from '../../_files';

import {
    ComponentFileSummary,
} from '../../../_utils';

> Filters are a way to automatically transform content before it is output. For example:

Reviewer: ```suggestion
Filters are a way to automatically transform content before it is output. Filters may be used to:
```

- Render embedded equations to images (the TeX filter)
- Links to media files can be automatically converted to an embedded players.
- Mentions of glossary terms can be automatically converted to links.

Filters are one of the easiest types of plugin to create.

> 

Reviewer: ```suggestion

Filters are applied to content passing in to the `format_string()` and `format_text()` functions, which are part of the [Output API](https://docs.moodle.org/dev/Output_functions).
```

Author: Thanks, a commit the suggestion.

<ComponentFileSummary
    required
    filepath="/filter.php"
    summary="Filter main class"
    plugintype="filter"
    pluginname="pluginname"
    example={Filter}
    description="The filter file contains the code for the main filter file. Unlike more complex plguins like activities or repositories, filters only have one mandatory class extending the core moodle_text_filter class."
/>

### version.php

<VersionPHP
    plugintype="filter"
/>

### lang/en/filter_pluginname.php

<!-- markdownlint-disable-next-line MD038 -->
export const langExample = `
 $string['filtername'] = 'Activity names auto-linking';
`;

<Lang
    plugintype="filter"
    pluginname="pluginname"
    example={langExample}
/>

## Test a filter

To enable a filter go to the [filters administration screen](./) and set the filter active to "On".

Filters are applied to all text that is printed with the output functions `format_text()` and `format_string()`. To see a filter in action, add some content to a label resource. When you look at that course in the course listing, you should see that your filter has transformed the text accordingly.

## Filter performance

> Is important to note that all active filters will be called to transform every bit of text output using `format_text()`, and possibly also `format_string()`. That means that a filter plugin could cause big performance problems. It is extremely important to uses caches when a filter have to get data out of the database,

Reviewer: ```suggestion
It is important to note that all active filters will be called to transform every bit of text output using `format_text()` (headers and content), and `format_string()` (headers only). As a result a filter plugin can cause big performance problems. It is extremely important to uses caches if your filter must retrieve data from the database, or other similar sources.
```

If a filter uses a special syntax or it is based on an appearance of a substring in the text, it is recommend to perform a quick and cheap `strpos()` search first prior to executing the full regex-based search and replace.

<details>
  <summary>View example</summary>
  <div>

```php
/**
 * Example of a filter that uses <a> links in some way.
 */
public function filter($text, array $options = []) {

    if (!is_string($text) or empty($text)) {
        // Non-string data can not be filtered anyway.
        return $text;
    }

    if (stripos($text, '</a>') === false) {
        // Performance shortcut - if there is no </a> tag, nothing can match.
        return $text;
    }

    // Here we can perform some more complex operations with the <a>
    // links in the text.
}
```

  </div>
</details>

## Local configuration

> In addition filters can also have different configuration in each context.
> 
> For example, the glossary filter could be changed so that in Forum A, you can choose to only link words from a particular glossary, say Glossary A, while in Forum B you choose to link words from Glossary B.

Reviewer: ```suggestion
Filters can use different configuration depending on the context in which they are called. For example, the glossary filter can be configured such that when displayed in Forum A it only links words from a particular glossary, while in Forum B it links words from a different glossary..
```

Author: Thanks, a commit the suggestion.

> To do that sort of thing, the filter plugin must provide a **filterlocalsettings.php** file. In it, you must define a Moodle form that is a subclass of filter_local_settings_form. In addition to the standard formslib methods, you also need to define a save_changes method.

Reviewer: ```suggestion
To support this behaviour, a filter plugin must provide a `filterlocalsettings.php` file which defines a Moodle form which subclasses the `filter_local_settings_form` class. In addition to the standard formslib methods, you must also define a `save_changes` method.
```
